pymc4/gp/gp.py

```diff
@@ -0,0 +1,204 @@
+import tensorflow as tf
+from .mean import Zero
+from ..distributions import MvNormal, Normal
+from .util import stabilize
+
+
+__all__ = ["LatentGP"]
+
+
+class BaseGP:
+    def __init__(self, cov_fn, mean_fn=Zero(1)):
+        if mean_fn.feature_ndims != cov_fn.feature_ndims:
+            raise ValueError("The feature_ndims of mean and covariance functions should be equal")
+        self.feature_ndims = mean_fn.feature_ndims
+        self.mean_fn = mean_fn
+        self.cov_fn = cov_fn
+
+    def prior(self, name, X, **kwargs):
+        raise NotImplementedError
+
+    def conditional(self, name, Xnew, **kwargs):
+        raise NotImplementedError
+
+    def predict(self, name, Xnew, **kwargs):
+        raise NotImplementedError
+
+    def marginal_likelihood(self, name, X, **kwargs):
+        raise NotImplementedError
+
+
+class LatentGP(BaseGP):
+    r"""Latent Gaussian process.
+
+    The `gp.LatentGP` class is a direct implementation of a GP.  No additive noise is assumed.
+    It is called "Latent" because the underlying function values are treated as latent variables.
+    It has a `prior` method and a `conditional` method.  Given a mean and covariance function the
+    function :math:`f(x)` is modeled as,
+
+    .. math::
+       f(x) \sim \mathcal{GP}\left(\mu(x), k(x, x')\right)
+
+    Use the `prior` and `conditional` methods to actually construct random variables representing
+    the unknown, or latent, function whose distribution is the GP prior or GP conditional.
+    This GP implementation can be used to implement regression on data that is not normally
+    distributed. For more information on the `prior` and `conditional` methods,
+    see their docstrings.
+
+    Parameters
+    ----------
+    cov_fn : pm.gp.Covariance
+        The covariance function.
+    mean_fn : pm.gp.Mean
+        The mean function.
+
+    Examples
+    --------
+    We can construct a LatentGP class with multiple ``feature_ndims`` and multiple batches.
+    Below is an example with ``feature_ndims=2`` , ``batch_shape=2``, 10 prior samples,
+    and 5 new samples. Notice that unlike PyMC3, ``given`` in ``conditional`` method is
+    NOT optional.
+    
+    .. code:: python
+
+        X = np.random.randn(2, 10, 2, 2)
+        Xnew = np.random.randn(2, 5, 2, 2)
+
+        # Let's define out GP model and its parameters
+        mean_fn = pm.gp.mean.Zero(feature_ndims=2)
+        cov_fn = pm.gp.cov.ExpQuad(1., 1., feature_ndims=2)
+        gp = pm.gp.LatentGP(mean_fn, cov_fn)
+
+        @pm.model
+        def gpmodel():
+            prior = yield gp.prior('prior', X)
+            cond = yield gp.conditional('cond', Xnew, given={'X': X, 'f': prior})
```

Comment:
I'm a bit confused by conditioning on the prior?  For prediction you condition on the posterior.

Reply:
Yeah! I called it ``prior`` in a sense that we are using the gp prior to infer the conditional. Will change it though as the names ``f`` and ``fcond`` sound more intuitive

Reply:
Ah I see now, it does work the same way.  The naming threw me off.  Agree w the change